docs/DesignDocs/enclave_openssl_support.md

```diff
@@ -0,0 +1,90 @@
+  - This sample serves as the analog to the existing mbedTLS-specific file-encryptor sample.
+- Integrate an OpenSSL variant into existing OE SDK samples that rely on crypto:
+  - [attested_tls](https://github.com/openenclave/openenclave-openssl/tree/master/sample/attested_tls)
+  - data-sealing
+  - local_attestation
+  - remote_attestation
+- Document the support for OpenSSL in OE SDK, i.e. add documentation analogous to [MbedtlsSupport.md](https://github.com/openenclave/openenclave/blob/master/docs/MbedtlsSupport.md)
+
+Also, all prototypes have technical gaps that will need to be addressed for production purposes. A non-exhaustive list of such tasks would include:
+
+- [ ] Add ctr_drbg.c to the `oecryptoopenssl` library implementation for parity with `oecryptombedtls`.
+- [ ] Update OpenSSL code changes to use `oe_get_entropy()` for entropy pool seeding, and avoid modifying the OpenSSL drbg code.
+- [ ] Formalize additional syscall stubs needed for OpenSSL as needed for [ssl_stubs.c](https://github.com/openenclave/openenclave-openssl/blob/master/0001-Get-openssl-to-build-against-MUSL-headers.-Use-SGX-r.patch).
+- [ ] Update the OpenSSL target branch to the latest [OpenSSL_1_1_1-stable](https://github.com/openssl/openssl/tree/OpenSSL_1_1_1-stable) and resolve test coverage differences or breaking changes.
+- [ ] Evaluate the set of [broken OpenSSL tests](https://github.com/openenclave/openenclave-openssl/blob/master/test/tests.broken) when executed in an enclave.
+
+## Alternates
+
+As mentioned in the motivation, there is a desire for a long-term solution to enable OE SDK users to replace the crypto library used by the OE enclave runtime with an arbitrary crypto library. One approach to enable this is to formalize the internal OE crypto API that is being implemented by `oecryptombedtls` and `oecryptoopenssl` in this proposal. This approach is an additive change that we may pursue in the future and does not obviate the support proposed here.
+
+An additional refinement of that approach is to eliminate the internal OE crypto API layer entirely:
+
+- OE SDK calls OpenSSL APIs directly instead of using its `oecrypto` indirection.
+- `oecryptoopenssl` becomes unnecessary, and `oecryptombedtls` would need to be refactored to be an mbedTLS to OpenSSL API adaptor instead.
+- Other crypto libraries that are already compatible or have existing shims with the OpenSSL API can be reused with minimal change.
+
+This option is not being pursued currently because:
+
+- As with the option to expose the existing `oecrypto` API for adaptation, it is additive to the proposed solution and can be incrementally pursued at a later date.
+  - Since it doesn't formalize the `oecrypto` API for crypto libraries to adapt against publicly at this time, the current proposal also avoids introducing a breaking change for enclave crypto libraries in the future.
```

Comment:
If we want to break them eventually, it's almost always better to break sooner, when there's less stuff to break.
However, another approach to avoid breaking, is to provide an optional shim of "deprecated" oecrypto APIs over the actual ones, but that just depends on how long we want to support the "deprecated" APIs.
I'd be in favor of saying the north star is to deprecate the oecrypto API, even if it's not the first step.

Reply:
The point is that the proposed design sidesteps the question of deprecating or breaking APIs entirely, by avoiding making the oecrypto API public to begin with. As it remains an internal API, whether it is eventually removed or not does not involve a public deprecation strategy.